llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp

```diff
@@ -3666,6 +3666,63 @@ LegalizerHelper::bitcastConcatVector(MachineInstr &MI, unsigned TypeIdx,
   return Legalized;
 }
 
+/// This attempts to bitcast G_EXTRACT_SUBVECTOR to CastTy.
+///
+///  <vscale x 8 x i1> = G_EXTRACT_SUBVECTOR <vscale x 16 x i1>, N
+///
+/// ===>
+///
+///  <vscale x 2 x i1> = G_BITCAST <vscale x 16 x i1>
+///  <vscale x 1 x i8> = G_EXTRACT_SUBVECTOR <vscale x 2 x i1>, N / 8
+///  <vscale x 8 x i1> = G_BITCAST <vscale x 1 x i8>
+LegalizerHelper::LegalizeResult
+LegalizerHelper::bitcastExtractSubvector(MachineInstr &MI, unsigned TypeIdx,
+                                         LLT CastTy) {
+  auto ES = dyn_cast<GExtractSubvector>(&MI);
+  if (!ES)
+    return UnableToLegalize;
+
+  if (!CastTy.isVector())
+    return UnableToLegalize;
+
+  if (TypeIdx != 0)
+    return UnableToLegalize;
+
+  Register Dst = ES->getReg(0);
+  Register Src = ES->getSrcVec();
+  uint64_t Idx = ES->getIndexImm();
+
+  MachineRegisterInfo &MRI = *MIRBuilder.getMRI();
+
+  LLT DstTy = MRI.getType(Dst);
+  LLT SrcTy = MRI.getType(Src);
+  ElementCount DstTyEC = DstTy.getElementCount();
+  ElementCount SrcTyEC = SrcTy.getElementCount();
+  auto DstTyMinElts = DstTyEC.getKnownMinValue();
+  auto SrcTyMinElts = SrcTyEC.getKnownMinValue();
+
+  if (DstTy == CastTy)
+    return Legalized;
+
+  auto AdjustAmt = CastTy.getScalarType().getSizeInBits();
+  if (DstTyMinElts < AdjustAmt || SrcTyMinElts < AdjustAmt)
```

Comment:
Is this redundant with the `DstTyMinElts % AdjustAmt != 0` and `SrcTyMinElts % AdjustAmt != 0` checks?

Reply:
In selectiondag:

```
if (VecVT.getVectorMinNumElements() >= 8 && SubVecVT.getVectorMinNumElements() >= 8) {
  assert(OrigIdx % 8 == 0 && "Invalid index");
  assert(VecVT.getVectorMinNumElements() % 8 == 0 && SubVecVT.getVectorMinNumElements() % 8 == 0 && "Unexpected mask vector lowering");
  ...
}
```

I was bringing this over. It's probably redundant in both cases?

Reply:
Your AdjustAmt isn't correct. It only works if the source type is i1. For any other type you need the ratio of the new and old element sizes.

Reply:
> In selectiondag:
> 
> ```
> if (VecVT.getVectorMinNumElements() >= 8 && SubVecVT.getVectorMinNumElements() >= 8) {
>   assert(OrigIdx % 8 == 0 && "Invalid index");
>   assert(VecVT.getVectorMinNumElements() % 8 == 0 && SubVecVT.getVectorMinNumElements() % 8 == 0 && "Unexpected mask vector lowering");
>   ...
> }
> ```
> 
> I was bringing this over. It's probably redundant in both cases?

The SelectionDAG asserts checking something more specific. Because that code runs after legalization, it should only receive a legal type. The only legal types with more than 8 elements have a multiple of 8 elements.

The GISel code doesn't get to deal with only legal types.

Reply:
removed.